### DISCUSSION TITLE
Access Context Variables in JavaScript

### DISCUSSION THREAD

**Discussion by swainn** [tethysplatform/tethys#678] [Tips and Tricks]
At some point when you are developing, you will likely need access to context variables in your JavaScript. The context variables are not made accessible to the JavaScript automatically. So, I like to use the "data attribute" pattern to do this. It works something like this:

1. Include the variable in your context object in the controller:

```python
def my_controller(request):
    ...
    context = {
        'some_variable': 'some_value',
    }
    return render(request, 'some_template.html', context)
```

2. Create a `div` element in your HTML template for storing attributes you'd like to access in JavaScript. Give the element an ID so it is easy to find/select in the JavaScript. Then, add a `data` attribute for each context variable you'd like access to in the JavaScript using the pattern `data-<variable-name>`:

```django
<div id="my-attributes" data-some-variable="{{ some_variable }}"></div>
```

3. In your JavaScript, select the element by its ID and retrieve the data attribute values. The JQuery (`$`) `data()` method makes this very easy:

```javascript
// Note: when using the $.data method, leave off the "data" part of the attribute name
var some_variable = $('#my-attributes').data('some-variable');
```

See: [jQuery.data() documentation](https://api.jquery.com/jquery.data/) for more details.

**Comment by rileyhales** [comment]
You can also create a `<script>` inside the template being rendered. This might look like:

```
<script>
const myVarName = {{ some_variable }}
</script>
```

Note: {{ some_variable }} indicates that django will render the context variable named some_variable in place of {{ some_variable }}. Place this script tag before the other script tags which are pulling the `.js` files or mark those as defer so they won't run until after this tag. depending on the type of some_variable, you might need additional syntax e.g.

```
const myVarName = JSON.parse("{{ some_json_dumped_to_a_string }}")
const myVarName = "{{ some_string_variable }}"
const myVarName = {{ some_int_or_float_variable }}
```
